File: C#/TextToImageHandler.cs
using System;
using System.Net;
using System.Threading.Tasks;
using System.IO;
using System.Text;
using System.Net.Http;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Linq;
using System.Threading;

namespace ChatOnServer
{
    public class TextToImageHandler
    {
        private readonly HttpClient httpClient = new HttpClient();
        private readonly SemaphoreSlim semaphore = new SemaphoreSlim(1, 1); // 用于同步文件操作

        public async Task Handle(HttpListenerContext context)
        {
            var request = context.Request;
            var response = context.Response;

            try
            {
                // 设置 CORS 头
                response.Headers.Add("Access-Control-Allow-Origin", "*");
                response.Headers.Add("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
                response.Headers.Add("Access-Control-Allow-Headers", "Content-Type, Authorization");

                // 处理预检请求
                if (request.HttpMethod.Equals("OPTIONS", StringComparison.OrdinalIgnoreCase))
                {
                    response.StatusCode = 204; // No Content
                    response.Close();
                    return;
                }

                // 只允许 POST 请求
                if (!request.HttpMethod.Equals("POST", StringComparison.OrdinalIgnoreCase))
                {
                    response.StatusCode = 405; // Method Not Allowed
                    byte[] methodNotAllowedBytes = Encoding.UTF8.GetBytes("Method Not Allowed");
                    await response.OutputStream.WriteAsync(methodNotAllowedBytes, 0, methodNotAllowedBytes.Length);
                    response.Close();
                    return;
                }

                // 读取请求体
                string requestBody;
                using (var reader = new StreamReader(request.InputStream, request.ContentEncoding))
                {
                    requestBody = await reader.ReadToEndAsync();
                }

                Console.WriteLine($"Received Image Generations JSON: {requestBody}");

                // 解析 JSON 请求
                JsonDocument requestJson;
                try
                {
                    requestJson = JsonDocument.Parse(requestBody);
                }
                catch (JsonException je)
                {
                    Utils.SendError(response, $"JSON 解析错误: {je.Message}");
                    return;
                }

                JsonElement root = requestJson.RootElement;

                // 验证 'prompt' 字段是否存在
                if (!root.TryGetProperty("prompt", out JsonElement promptElem))
                {
                    Utils.SendError(response, "缺少必需的字段: prompt");
                    return;
                }

                string prompt = promptElem.GetString()?.Trim() ?? string.Empty;
                if (string.IsNullOrEmpty(prompt))
                {
                    Utils.SendError(response, "Prompt 不能为空。");
                    return;
                }

                // 处理 'response_format'，默认值为 'b64_json'
                string responseFormat = root.TryGetProperty("response_format", out JsonElement formatElem)
                    ? formatElem.GetString()?.Trim() ?? "b64_json"
                    : "b64_json";

                Console.WriteLine($"Prompt: {prompt}");
                Console.WriteLine($"Response Format: {responseFormat}");

                // 构建新的 JSON 负载
                var newRequest = new
                {
                    image_aspect_ratio = "1:1",
                    function_image_gen = true,
                    max_tokens = 8000,
                    function_web_search = true,
                    messages = new[]
                    {
                        new { role = "system", content = "You are a helpful artist, please draw a picture.Based on imagination, draw a picture with user message." },
                        new { role = "user", content = $"Draw: {prompt}" }
                    },
                    model = "gpt-4o",
                    image_style = "anime",
                    source = "chat/pro_image"
                };

                string modifiedRequestBody = JsonSerializer.Serialize(newRequest);
                Console.WriteLine($"构建的请求 JSON: {modifiedRequestBody}");

                // 生成 Bearer Token
                string[] tmpToken = BearerTokenGenerator.GetBearer(modifiedRequestBody);
                if (tmpToken == null || tmpToken.Length == 0)
                {
                    Utils.SendError(response, "Bearer Token 生成失败。");
                    return;
                }

                // 构建 HTTP 请求
                var apiRequest = Utils.BuildHttpRequest(modifiedRequestBody, tmpToken);

                // 发送请求并处理 SSE 流
                using (var apiResponse = await httpClient.SendAsync(apiRequest, HttpCompletionOption.ResponseHeadersRead))
                {
                    if (!apiResponse.IsSuccessStatusCode)
                    {
                        Utils.SendError(response, $"API 错误: {apiResponse.StatusCode}");
                        return;
                    }

                    // 读取响应流
                    using (var stream = await apiResponse.Content.ReadAsStreamAsync())
                    using (var reader = new StreamReader(stream))
                    {
                        StringBuilder urlBuilder = new StringBuilder();

                        string line;
                        while ((line = await reader.ReadLineAsync()) != null)
                        {
                            if (line.StartsWith("data: "))
                            {
                                string data = line.Substring(6).Trim();
                                if (data.Equals("[DONE]", StringComparison.OrdinalIgnoreCase))
                                {
                                    break; // SSE 流结束
                                }

                                try
                                {
                                    JsonDocument sseJson = JsonDocument.Parse(data);
                                    JsonElement sseRoot = sseJson.RootElement;

                                    if (sseRoot.TryGetProperty("choices", out JsonElement choices))
                                    {
                                        foreach (JsonElement choice in choices.EnumerateArray())
                                        {
                                            if (choice.TryGetProperty("delta", out JsonElement delta))
                                            {
                                                if (delta.TryGetProperty("content", out JsonElement contentElem))
                                                {
                                                    string content = contentElem.GetString();
                                                    urlBuilder.Append(content);
                                                }
                                            }
                                        }
                                    }
                                }
                                catch (JsonException je)
                                {
                                    Console.WriteLine($"SSE 中的 JSON 解析错误: {je.Message}");
                                }
                            }
                        }

                        string imageMarkdown = urlBuilder.ToString();
                        Console.WriteLine($"Image Markdown: {imageMarkdown}");

                        // 验证构建的 Markdown
                        if (string.IsNullOrEmpty(imageMarkdown))
                        {
                            Utils.SendError(response, "无法从 SSE 流中构建图像 Markdown。");
                            return;
                        }

                        // 使用正则表达式提取图像路径
                        string extractedPath = ExtractPathFromMarkdown(imageMarkdown);
                        if (string.IsNullOrEmpty(extractedPath))
                        {
                            Utils.SendError(response, "无法从 Markdown 中提取图像路径。");
                            return;
                        }

                        Console.WriteLine($"提取的路径: {extractedPath}");

                        string prefix = "https://spc.unk/";
                        if (extractedPath.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
                        {
                            extractedPath = extractedPath.Substring(prefix.Length);
                        }

                        Console.WriteLine($"过滤后的路径: {extractedPath}");

                        // 构建存储 URL
                        string storageUrl = $"https://api.chaton.ai/storage/{extractedPath}";
                        Console.WriteLine($"存储 URL: {storageUrl}");

                        // 获取最终下载 URL
                        string finalDownloadUrl = await FetchGetUrlFromStorageAsync(storageUrl);
                        if (string.IsNullOrEmpty(finalDownloadUrl))
                        {
                            Utils.SendError(response, "无法从存储中获取最终下载 URL。");
                            return;
                        }

                        Console.WriteLine($"最终下载 URL: {finalDownloadUrl}");

                        // 下载图像
                        byte[] imageBytes = await DownloadImageAsync(finalDownloadUrl);
                        if (imageBytes == null)
                        {
                            Utils.SendError(response, "无法从最终 URL 下载图像。");
                            return;
                        }

                        // 将图像编码为 Base64
                        string imageBase64 = Convert.ToBase64String(imageBytes);

                        // 根据请求的格式构建响应
                        if (responseFormat.Equals("b64_json", StringComparison.OrdinalIgnoreCase))
                        {
                            var responseJson = new
                            {
                                data = new[]
                                {
                                    new { b64_json = imageBase64 }
                                }
                            };

                            string responseBody = JsonSerializer.Serialize(responseJson);
                            byte[] responseBytes = Encoding.UTF8.GetBytes(responseBody);

                            response.ContentType = "application/json; charset=UTF-8";
                            response.ContentLength64 = responseBytes.Length;
                            response.StatusCode = 200;
                            await response.OutputStream.WriteAsync(responseBytes, 0, responseBytes.Length);
                            response.OutputStream.Close(); // 在这里关闭响应
                        }
                        else
                        {
                            Utils.SendError(response, $"不支持的 response_format: {responseFormat}");
                            return;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"内部服务器错误: {ex.Message}");
                Utils.SendError(response, $"内部服务器错误: {ex.Message}");
            }
        }

        /// <summary>
        /// 使用正则表达式从 Markdown 字符串中提取图像路径。
        /// </summary>
        private string ExtractPathFromMarkdown(string markdown)
        {
            // 匹配 ![Image](URL) 的正则表达式
            Regex regex = new Regex(@"!\[.*?\]\((.*?)\)");
            Match match = regex.Match(markdown);

            if (match.Success && match.Groups.Count > 1)
            {
                return match.Groups[1].Value;
            }

            return null;
        }

        /// <summary>
        /// 从存储 URL 的 JSON 响应中获取 'getUrl'。
        /// </summary>
        private async Task<string> FetchGetUrlFromStorageAsync(string storageUrl)
        {
            try
            {
                HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, storageUrl);
                HttpResponseMessage response = await httpClient.SendAsync(request);

                if (response.StatusCode != HttpStatusCode.OK)
                {
                    Console.WriteLine($"获取存储 URL 失败。状态码: {response.StatusCode}");
                    return null;
                }

                string responseBody = await response.Content.ReadAsStringAsync();
                JsonDocument jsonResponse = JsonDocument.Parse(responseBody);
                JsonElement root = jsonResponse.RootElement;

                if (root.TryGetProperty("getUrl", out JsonElement getUrlElem))
                {
                    return getUrlElem.GetString();
                }
                else
                {
                    Console.WriteLine("JSON 响应中不包含 'getUrl' 字段。");
                    return null;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"从存储 URL 获取 'getUrl' 时出错: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// 从指定的 URL 下载图像。
        /// </summary>
        private async Task<byte[]> DownloadImageAsync(string imageUrl)
        {
            try
            {
                HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, imageUrl);
                HttpResponseMessage response = await httpClient.SendAsync(request);

                if (response.StatusCode != HttpStatusCode.OK)
                {
                    Console.WriteLine($"下载图像失败。状态码: {response.StatusCode}");
                    return null;
                }

                return await response.Content.ReadAsByteArrayAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"下载图像时出错: {ex.Message}");
                return null;
            }
        }
    }
}

File: java/TextToImageHandler.java
import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.*;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;
import utils.utils;
import utils.BearerTokenGenerator;

public class TextToImageHandler implements HttpHandler {
    private final HttpClient httpClient = HttpClient.newHttpClient();
    private final Executor executor = Executors.newSingleThreadExecutor();

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        // 设置 CORS 头
        Headers headers = exchange.getResponseHeaders();
        headers.add("Access-Control-Allow-Origin", "*");
        headers.add("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
        headers.add("Access-Control-Allow-Headers", "Content-Type, Authorization");

        String requestMethod = exchange.getRequestMethod().toUpperCase();

        // 处理预检请求
        if (requestMethod.equals("OPTIONS")) {
            exchange.sendResponseHeaders(204, -1);
            return;
        }

        // 只允许 POST 请求
        if (!"POST".equals(requestMethod)) {
            exchange.sendResponseHeaders(405, -1);
            return;
        }

        // 异步处理请求
        CompletableFuture.runAsync(() -> {
            try {
                // 读取请求体
                InputStream is = exchange.getRequestBody();
                String requestBody = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))
                        .lines()
                        .reduce("", (acc, line) -> acc + line);

                System.out.println("Received Image Generations JSON: " + requestBody);

                JSONObject userInput = new JSONObject(requestBody);

                // 验证必需的字段
                if (!userInput.has("prompt")) {
                    utils.sendError(exchange, "缺少必需的字段: prompt");
                    return;
                }

                String userPrompt = userInput.optString("prompt", "").trim();
                String responseFormat = userInput.optString("response_format", "").trim();
                int n = userInput.optInt("n", 1); // 读取 n 的值，默认为 1

                if (userPrompt.isEmpty()) {
                    utils.sendError(exchange, "Prompt 不能为空。");
                    return;
                }

                System.out.println("Prompt: " + userPrompt);
                System.out.println("Number of images to generate (n): " + n);

                // 设置最大尝试次数为 2 * n
                int maxAttempts = 2 * n;
                System.out.println("Max Attempts: " + maxAttempts);

                // 初始化用于存储多个 URL 的列表
                List<String> finalDownloadUrls = new ArrayList<>();

                // 开始尝试生成图像
                attemptGenerateImages(userPrompt, responseFormat, n, maxAttempts, 1, finalDownloadUrls, exchange)
                        .thenAccept(success -> {
                            if (success) {
                                // 根据 response_format 返回相应的响应
                                boolean isBase64Response = "b64_json".equalsIgnoreCase(responseFormat);

                                JSONObject responseJson = new JSONObject();
                                responseJson.put("created", System.currentTimeMillis() / 1000); // 添加 created 字段
                                JSONArray dataArray = new JSONArray();

                                if (isBase64Response) {
                                    // 对每个下载链接进行处理
                                    for (String downloadUrl : finalDownloadUrls) {
                                        try {
                                            // 下载图像并编码为 Base64
                                            byte[] imageBytes = downloadImage(downloadUrl);
                                            if (imageBytes == null) {
                                                // 如果下载失败，跳过此链接
                                                System.err.println("无法从 URL 下载图像: " + downloadUrl);
                                                continue;
                                            }

                                            // 转换为 Base64
                                            String imageBase64 = Base64.getEncoder().encodeToString(imageBytes);

                                            JSONObject dataObject = new JSONObject();
                                            dataObject.put("b64_json", imageBase64);
                                            dataArray.put(dataObject);
                                        } catch (Exception e) {
                                            e.printStackTrace();
                                            // 继续处理其他图像
                                        }
                                    }
                                } else {
                                    // 直接返回所有图像的 URL
                                    for (String downloadUrl : finalDownloadUrls) {
                                        JSONObject dataObject = new JSONObject();
                                        dataObject.put("url", downloadUrl);
                                        dataArray.put(dataObject);
                                    }
                                }

                                // 如果收集的 URL 数量不足 n，则通过复制现有的 URL 来填充
                                while (dataArray.length() < n && dataArray.length() > 0) {
                                    for (int i = 0; i < dataArray.length() && dataArray.length() < n; i++) {
                                        JSONObject original = dataArray.getJSONObject(i);
                                        dataArray.put(original);
                                    }
                                }

                                responseJson.put("data", dataArray);

                                try {
                                    byte[] responseBytes = responseJson.toString().getBytes(StandardCharsets.UTF_8);
                                    exchange.getResponseHeaders().add("Content-Type", "application/json");
                                    exchange.sendResponseHeaders(200, responseBytes.length);
                                    try (OutputStream os = exchange.getResponseBody()) {
                                        os.write(responseBytes);
                                    }
                                } catch (IOException e) {
                                    e.printStackTrace();
                                }

                            } else {
                                // 如果在所有尝试后仍未收集到足够的链接，则返回错误
                                utils.sendError(exchange, "无法生成足够数量的图像。");
                            }
                        })
                        .exceptionally(ex -> {
                            ex.printStackTrace();
                            utils.sendError(exchange, "请求失败: " + ex.getMessage());
                            return null;
                        });

            } catch (JSONException je) {
                je.printStackTrace();
                utils.sendError(exchange, "JSON 解析错误: " + je.getMessage());
            } catch (Exception e) {
                e.printStackTrace();
                utils.sendError(exchange, "内部服务器错误: " + e.getMessage());
            }
        }, executor);
    }

    /**
     * 尝试生成图像，带有重试机制。
     *
     * @param userPrompt       用户的提示
     * @param responseFormat   响应格式
     * @param n                需要生成的图像数量
     * @param maxAttempts      最大尝试次数
     * @param currentAttempt   当前尝试次数
     * @param finalDownloadUrls 收集的最终下载链接列表
     * @param exchange         HttpExchange 对象
     * @return CompletableFuture<Boolean> 表示是否成功收集到足够的下载链接
     */
    private CompletableFuture<Boolean> attemptGenerateImages(String userPrompt, String responseFormat, int n,
                                                             int maxAttempts, int currentAttempt,
                                                             List<String> finalDownloadUrls, HttpExchange exchange) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                // 构建新的 TextToImage JSON 请求体
                JSONObject textToImageJson = new JSONObject();
                textToImageJson.put("function_image_gen", true);
                textToImageJson.put("function_web_search", true);
                textToImageJson.put("image_aspect_ratio", "1:1");
                textToImageJson.put("image_style", "anime"); // 固定 image_style
                textToImageJson.put("max_tokens", 8000);
                textToImageJson.put("n", 1); // 每次请求生成一张图像
                JSONArray messages = new JSONArray();
                JSONObject message = new JSONObject();
                message.put("content", "You are a helpful artist, please draw a picture. Based on imagination, draw a picture with user message.");
                message.put("role", "system");
                JSONObject userMessage = new JSONObject();
                userMessage.put("content", "Draw: " + userPrompt);
                userMessage.put("role", "user");
                messages.put(message);
                messages.put(userMessage);
                textToImageJson.put("messages", messages);
                textToImageJson.put("model", "gpt-4o"); // 固定 model
                textToImageJson.put("source", "chat/pro_image"); // 固定 source

                String modifiedRequestBody = textToImageJson.toString();

                // 构建请求
                String[] tmpToken = BearerTokenGenerator.GetBearer(modifiedRequestBody);
                System.out.println("Attempt " + currentAttempt + " - 构建的请求: " + modifiedRequestBody);
                HttpRequest request = utils.buildHttpRequest(modifiedRequestBody, tmpToken);

                // 发送请求并处理 SSE 流
                HttpResponse.BodyHandler<Stream<String>> bodyHandler = HttpResponse.BodyHandlers.ofLines();
                HttpResponse<Stream<String>> response = httpClient.send(request, bodyHandler);

                if (response.statusCode() != 200) {
                    System.err.println("Attempt " + currentAttempt + " - API 错误: " + response.statusCode());
                    return false;
                }

                // 初始化用于拼接 URL 的 StringBuilder
                StringBuilder urlBuilder = new StringBuilder();

                // 读取 SSE 流并拼接 URL
                response.body().forEach(line -> {
                    if (line.startsWith("data: ")) {
                        String data = line.substring(6).trim();
                        if (data.equals("[DONE]")) {
                            return; // 完成读取
                        }

                        try {
                            JSONObject sseJson = new JSONObject(data);
                            if (sseJson.has("choices")) {
                                JSONArray choices = sseJson.getJSONArray("choices");
                                for (int i = 0; i < choices.length(); i++) {
                                    JSONObject choice = choices.getJSONObject(i);
                                    JSONObject delta = choice.optJSONObject("delta");
                                    if (delta != null && delta.has("content")) {
                                        String content = delta.getString("content");
                                        urlBuilder.append(content);
                                    }
                                }
                            }
                        } catch (JSONException e) {
                            System.err.println("Attempt " + currentAttempt + " - JSON解析错误: " + e.getMessage());
                        }
                    }
                });

                String imageMarkdown = urlBuilder.toString();
                // Step 1: 检查Markdown文本是否为空
                if (imageMarkdown.isEmpty()) {
                    System.out.println("Attempt " + currentAttempt + " - 无法从 SSE 流中构建图像 Markdown。");
                    return false;
                }

                // Step 2: 从Markdown中提取图像路径
                String extractedPath = extractPathFromMarkdown(imageMarkdown);

                // Step 3: 如果没有提取到路径，输出错误信息
                if (extractedPath == null || extractedPath.isEmpty()) {
                    System.out.println("Attempt " + currentAttempt + " - 无法从 Markdown 中提取路径。");
                    return false;
                }

                // Step 4: 过滤掉 "https://spc.unk/" 前缀
                extractedPath = extractedPath.replace("https://spc.unk/", "");

                // 输出提取到的路径
                System.out.println("Attempt " + currentAttempt + " - 提取的路径: " + extractedPath);

                // Step 5: 拼接最终的存储URL
                String storageUrl = "https://api.chaton.ai/storage/" + extractedPath;
                System.out.println("Attempt " + currentAttempt + " - 存储URL: " + storageUrl);

                // 请求 storageUrl 获取 JSON 数据
                String finalDownloadUrl = fetchGetUrlFromStorage(storageUrl);
                if (finalDownloadUrl == null || finalDownloadUrl.isEmpty()) {
                    System.out.println("Attempt " + currentAttempt + " - 无法从 storage URL 获取最终下载链接。");
                    return false;
                }

                System.out.println("Attempt " + currentAttempt + " - Final Download URL: " + finalDownloadUrl);

                // 添加到收集的下载链接列表中
                synchronized (finalDownloadUrls) {
                    if (finalDownloadUrls.size() < n) {
                        finalDownloadUrls.add(finalDownloadUrl);
                        System.out.println("Attempt " + currentAttempt + " - 收集到的下载链接数量: " + finalDownloadUrls.size());
                    }
                }

                return finalDownloadUrls.size() >= n;

            } catch (Exception e) {
                e.printStackTrace();
                System.err.println("Attempt " + currentAttempt + " - 处理响应时发生错误: " + e.getMessage());
                return false;
            }
        }, executor).thenCompose(success -> {
            if (success) {
                return CompletableFuture.completedFuture(true);
            } else if (currentAttempt < maxAttempts) {
                System.out.println("Attempt " + currentAttempt + " - 未收集到足够的下载链接，开始重试...");
                return attemptGenerateImages(userPrompt, responseFormat, n, maxAttempts, currentAttempt + 1, finalDownloadUrls, exchange);
            } else {
                System.out.println("已达到最大尝试次数，无法生成足够数量的图像。");
                return CompletableFuture.completedFuture(false);
            }
        });
    }

    /**
     * 提取 Markdown 图片链接中的路径
     *
     * @param markdown 图片的 Markdown 语法字符串
     * @return 提取出的路径，如果失败则返回 null
     */
    public String extractPathFromMarkdown(String markdown) {
        // 正则表达式匹配 ![Image](URL)
        Pattern pattern = Pattern.compile("!\\[.*?\\]\\((.*?)\\)");
        Matcher matcher = pattern.matcher(markdown);

        // 如果找到了匹配的路径，返回第一个URL
        if (matcher.find()) {
            return matcher.group(1);
        }

        // 如果没有找到匹配，返回null
        return null;
    }

    /**
     * 从 storage URL 获取 JSON 并提取 getUrl
     *
     * @param storageUrl 拼接后的 storage URL
     * @return getUrl 的值，如果失败则返回 null
     */
    private String fetchGetUrlFromStorage(String storageUrl) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(storageUrl))
                    .GET()
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                System.err.println("获取 storage URL 失败，状态码: " + response.statusCode());
                return null;
            }

            JSONObject jsonResponse = new JSONObject(response.body());
            if (jsonResponse.has("getUrl")) {
                return jsonResponse.getString("getUrl");
            } else {
                System.err.println("JSON 响应中缺少 'getUrl' 字段。");
                return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * 下载图像
     *
     * @param imageUrl 图像的最终下载 URL
     * @return 图像的字节数组，失败时返回 null
     */
    private byte[] downloadImage(String imageUrl) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(imageUrl))
                    .GET()
                    .build();

            HttpResponse<byte[]> response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());

            if (response.statusCode() == 200) {
                return response.body();
            } else {
                System.err.println("下载图像失败，状态码: " + response.statusCode());
                return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}
